### ISSUE TITLE
Too many open files error for large clusters

### ISSUE THREAD

**Issue by satnam6502** [coreos/etcd#1959]
Perhaps related to issue #872 using version 0.4.6 with large-ish Kubernetes clusters I also get the too many open files error (I saw this with 750 node and 1,000 node clusters). Perhaps this is my fault or perhaps this is a bug? A socket not being closed, as suggested by Brendan Burns? Thank you!

```
[etcd] Dec 17 19:43:37.279 INFO      | oneofmany-master is starting a new cluster
[etcd] Dec 17 19:43:37.280 INFO      | etcd server [name oneofmany-master, listen on 0.0.0.0:4001, advertised url http://127.0.0.1:4001]
[etcd] Dec 17 19:43:37.280 INFO      | peer server [name oneofmany-master, listen on :7001, advertised url http://oneofmany-master:7001]
[etcd] Dec 17 19:43:37.280 INFO      | oneofmany-master starting in peer mode
[etcd] Dec 17 19:43:37.280 INFO      | oneofmany-master: state changed from 'initialized' to 'follower'.
[etcd] Dec 17 19:43:37.280 INFO      | oneofmany-master: state changed from 'follower' to 'leader'.
[etcd] Dec 17 19:43:37.280 INFO      | oneofmany-master: leader changed from '' to 'oneofmany-master'.
[etcd] Dec 17 21:06:55.721 INFO      | oneofmany-master: snapshot of 10003 events at index 10003 completed
[etcd] Dec 17 22:30:17.144 INFO      | oneofmany-master: snapshot of 10003 events at index 20006 completed
[etcd] Dec 17 23:53:38.585 INFO      | oneofmany-master: snapshot of 10003 events at index 30009 completed
[etcd] Dec 18 01:17:00.012 INFO      | oneofmany-master: snapshot of 10003 events at index 40012 completed
[etcd] Dec 18 02:29:48.430 INFO      | oneofmany-master: snapshot of 10001 events at index 50013 completed
2014/12/18 03:16:38 http: Accept error: accept tcp [::]:4001: too many open files; retrying in 5ms
2014/12/18 03:16:38 http: Accept error: accept tcp [::]:4001: too many open files; retrying in 10ms
2014/12/18 03:16:38 http: Accept error: accept tcp [::]:4001: too many open files; retrying in 20ms
2014/12/18 03:16:38 http: Accept error: accept tcp [::]:4001: too many open files; retrying in 40ms
2014/12/18 03:16:40 http: Accept error: accept tcp [::]:4001: too many open files; retrying in 5ms
```

**Comment by nguyenchiencong** [comment]
+1

**Comment by dchen1107** [comment]
@satnam Checked on the limit in our Container VM: 

$ /sbin/sysctl fs.file-max
fs.file-max = 376175

Per process's limit is 1024 though, we could increase it.

**Comment by alban** [comment]
The fs.file-max limit is global and containers cannot have a specific fs.file-max limit. Is it correct?

When the problem happens, it would be good to check file-nr as well:
$ cat /proc/sys/fs/file-nr

**Comment by theambient** [comment]
I experience this too on bare Ubuntu 14.04

**Comment by Little-elephant** [comment]
Also seen this on Ubuntu 14.04, on three node cluster

**Comment by kelseyhightower** [comment]
@theambient @Little-elephant @satnam6502 Can we redo this test with etcd 2.0.3?

**Comment by satnam6502** [comment]
Once I have emerged from the black hole of writing an e2e test for Kubernetes I can try again with etcd 2.0.3 although I suspect one of my better here might beat me to it.

**Comment by theambient** [comment]
@kelseyhightower problem vanished after updating python-etcd client dependency - urlib3.

seems that client (python-etcd via urrlib3) somehow wrong closed tcp connection and etcd did not handle it resulting in hanging tcp connection as far as I remember in CLOSE_WAIT.

new connections where establishing faster then died connections become actually closed  and those file descriptors number exceeded allowed per process value (1024). 

In a nutshell `\lambda > \mu`

that was my case.

**Comment by kelseyhightower** [comment]
@theambient Glad to hear it. 

@satnam6502 Let me know when you get a chance to test etcd 2.0.x, since reporting this issue we've release etcd 2.0.4 and a few changes on master that might help with etcd and idle connections.

**Comment by satnam6502** [comment]
Thanks -- I should emerge from a e2e testing quagmire shortly so I hope to return my attention to scalability shortly.

**Comment by gtonic** [comment]
Think I've got the same issue here, using etcd 2.0.5, Ubuntu 14.10.

Configured a 3 node cluster and would expect an additional node to join as a proxy. It starts up listening for client requests, however when I curl on it, I get this one:

```
2015/03/31 15:12:20 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 5ms
2015/03/31 15:12:20 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 10ms
2015/03/31 15:12:20 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 20ms
2015/03/31 15:12:20 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 40ms
2015/03/31 15:12:20 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 80ms
2015/03/31 15:12:20 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 160ms
2015/03/31 15:12:20 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 320ms
2015/03/31 15:12:20 etcdserver: could not get cluster response from http://10.0.0.132:2380: Get http://10.0.0.132:2380/members: dial tcp 10.0.0.132:2380: too many open files
2015/03/31 15:12:20 etcdserver: could not get cluster response from http://10.0.0.17:2380: Get http://10.0.0.17:2380/members: dial tcp 10.0.0.17:2380: too many open files
2015/03/31 15:12:20 etcdserver: could not get cluster response from http://10.0.0.212:2380: Get http://10.0.0.212:2380/members: dial tcp 10.0.0.212:2380: too many open files
2015/03/31 15:12:20 proxy: etcdserver: could not retrieve cluster information from the given urls
2015/03/31 15:12:21 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 640ms
2015/03/31 15:12:21 http: Accept error: accept tcp 127.0.0.1:2379: too many open files; retrying in 1s
```

**Comment by yichengq** [comment]
@gtonic 
In summary, this normally happens when the number of concurrent clients exceeds the upper limit of file descriptors, which results in that etcd cannot communicate with peers or open file.

The improvement/fix of the root cause is now tracked at #2477 .

Our first-step plan is to expose metrics about fd usage, and log when it is in high ratio.

**Comment by wojtek-t** [comment]
@fgrzadkowski

**Comment by bahbka** [comment]
Stucked with same problem while testing etcd at ubuntu 14.04, etcd 2.0.9
Configured 3 nodes + 1 proxy, curl once at proxy (no other clients)
lsof at proxy
```
...
etcd      19667 19671          root  998u     IPv4            2794667       0t0        TCP localhost:2379->localhost:36607 (ESTABLISHED)
etcd      19667 19671          root  999u     IPv4            2794669       0t0        TCP localhost:36608->localhost:2379 (ESTABLISHED)
etcd      19667 19671          root 1000u     IPv4            2794670       0t0        TCP localhost:2379->localhost:36608 (ESTABLISHED)
etcd      19667 19671          root 1001u     IPv4            2794672       0t0        TCP localhost:36609->localhost:2379 (ESTABLISHED)
etcd      19667 19671          root 1002u     IPv4            2794673       0t0        TCP localhost:2379->localhost:36609 (ESTABLISHED)
etcd      19667 19671          root 1003u     IPv4            2794675       0t0        TCP localhost:36610->localhost:2379 (ESTABLISHED)
...
# lsof | grep -c etcd
5151
```

**Comment by garo** [comment]
I'm running a three node etcd cluster which seems to leak file handles (and etcd process memory). When I do lsof for the process I can see ~17000 sockets where lsof can't identify protocol. Example:

````
COMMAND   PID USER   FD   TYPE DEVICE SIZE/OFF       NODE NAME
etcd    26574 root  cwd    DIR   0,57     4096         12 /etcd-v2.0.3-linux-amd64
etcd    26574 root  rtd    DIR   0,57     4096          2 /
etcd    26574 root  txt    REG   0,57  6717056         22 /etcd-v2.0.3-linux-amd64/etcd
etcd    26574 root    0r   CHR    1,3      0t0 2845068861 /dev/null
etcd    26574 root    1w  FIFO    0,8      0t0 2845068765 pipe
etcd    26574 root    2w  FIFO    0,8      0t0 2845068766 pipe
etcd    26574 root    3u  sock    0,7      0t0 2845068879 can't identify protocol
etcd    26574 root    4u  0000    0,9        0       6389 anon_inode
etcd    26574 root    5u  sock    0,7      0t0 2845068880 can't identify protocol
etcd    26574 root    6u  sock    0,7      0t0 2845068881 can't identify protocol
etcd    26574 root    7u  sock    0,7      0t0 2845672672 can't identify protocol
etcd    26574 root    8u  sock    0,7      0t0 2845659564 can't identify protocol
etcd    26574 root    9u  sock    0,7      0t0 2845653202 can't identify protocol
etcd    26574 root   10u  sock    0,7      0t0 2845656129 can't identify protocol
etcd    26574 root   11u  sock    0,7      0t0 2845655728 can't identify protocol
etcd    26574 root   12u  sock    0,7      0t0 2845655078 can't identify protocol
```

The etcd logs show a lot of lines like these (not sure if they are relevant, in no particular order):

```
etcdhttp: unexpected error: etcdserver: request timed out\n","stream":"stderr","time":"2015-04-22T06:09:24.474980135Z"}

{"log":"2015/04/22 06:09:26 sender: dropping MsgProp because maximal number 64 of sender buffer entries to http://etcd-3:2380/raft has been reached\n","stream":"stderr","time":"2015-04-22T06:09:26.058700265Z"}
```

I'm running 2.0.9

**Comment by akamalov** [comment]
Using <b>etcd 2.0.10</b>. Configured three-node cluster with fourth joining as a proxy. Getting errors related to <b>too many open files</b>:

```
-- Logs begin at Fri 2015-05-15 12:26:16 UTC, end at Mon 2015-05-18 16:52:27 UTC. --
May 18 16:45:52 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:52 proxy: marked endpoint http://0.0.0.0:2379 available
May 18 16:45:52 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:52 proxy: marked endpoint http://0.0.0.0:2379 available
May 18 16:45:52 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:52 proxy: marked endpoint http://0.0.0.0:2379 available
May 18 16:45:47 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:47 proxy: unable to get response from 3 endpoint(s)
May 18 16:45:47 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:47 proxy: marked endpoint http://0.0.0.0:2379 unavailable
May 18 16:45:47 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:47 proxy: failed to direct request to http://0.0.0.0:2379: dial tcp 0.0.0.0:2379: too many open files
May 18 16:45:47 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:47 proxy: marked endpoint http://0.0.0.0:2379 unavailable
May 18 16:45:47 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:47 proxy: failed to direct request to http://0.0.0.0:2379: dial tcp 0.0.0.0:2379: too many open files
May 18 16:45:47 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:47 proxy: marked endpoint http://0.0.0.0:2379 unavailable
May 18 16:45:47 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:47 proxy: failed to direct request to http://0.0.0.0:2379: dial tcp 0.0.0.0:2379: too many open files
May 18 16:45:47 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:47 http: Accept error: accept tcp [::]:2379: too many open files; retrying in 5ms
May 18 16:45:36 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:36 proxy: marked endpoint http://0.0.0.0:2379 available
May 18 16:45:36 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:36 proxy: marked endpoint http://0.0.0.0:2379 available
May 18 16:45:36 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:36 proxy: marked endpoint http://0.0.0.0:2379 available
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 http: Accept error: accept tcp [::]:2379: too many open files; retrying in 10ms
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 proxy: unable to get response from 3 endpoint(s)
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 proxy: marked endpoint http://0.0.0.0:2379 unavailable
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 proxy: failed to direct request to http://0.0.0.0:2379: dial tcp 0.0.0.0:2379: too many open files
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 proxy: marked endpoint http://0.0.0.0:2379 unavailable
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 proxy: failed to direct request to http://0.0.0.0:2379: dial tcp 0.0.0.0:2379: too many open files
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 proxy: marked endpoint http://0.0.0.0:2379 unavailable
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 proxy: failed to direct request to http://0.0.0.0:2379: dial tcp 0.0.0.0:2379: too many open files
May 18 16:45:31 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:45:31 http: Accept error: accept tcp [::]:2379: too many open files; retrying in 5ms
May 18 16:42:13 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:42:13 proxy: listening for client requests on 127.0.0.1:8080
May 18 16:42:13 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:42:13 proxy: listening for client requests on 0.0.0.0:2379
May 18 16:42:13 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:42:13 proxy: using peer urls [http://10.9.159.65:2380 http://10.9.159.66:2380 http://10.9.159.67:2380] from cluster file .//var/lib/et
May 18 16:42:13 wwserver197.alphalocal.net etcd2[708]: 2015/05/18 16:42:13 etcd: already initialized as proxy before, starting as etcd proxy...
May 18 16:42:13 wwserver197.alphalocal.net systemd[1]: Starting etcd 2...
```

Configuration for the proxy, pertaining to ETCD2 is the following:

```
 - name: download-etcd2.service
    enable: true
    command: start
    content: |
      [Unit]
      After=network-online.target
      Before=etcd2.service
      Description=Download etcd2 Binaries
      Documentation=https://github.com/coreos/etcd/
      Requires=network-online.target
      [Service]
      Environment=ETCD2_RELEASE_TARBALL=https://github.com/coreos/etcd/releases/download/v2.1.0-alpha.0/etcd-v2.1.0-alpha.0-linux-amd64.tar.gz
      ExecStartPre=/bin/mkdir -p /opt/bin
      ExecStart=/bin/bash -c "curl --silent --location $ETCD2_RELEASE_TARBALL | tar xzv -C /opt"
      ExecStartPost=/bin/ln -s /opt/etcd-v2.1.0-alpha.0-linux-amd64/etcd /opt/bin/etcd2
      ExecStartPost=/bin/ln -s /opt/etcd-v2.1.0-alpha.0-linux-amd64/etcdctl /opt/bin/etcdctl2
      RemainAfterExit=yes
      Type=oneshot
      [Install]
      WantedBy=multi-user.target
  - name: etcd2.service
    enable: true
    command: start
    content: |
      [Unit]
      After=download-etcd2.service
      Description=etcd 2
      Documentation=https://github.com/coreos/etcd/
      [Service]
      Environment="ETCD_NAME=wwserver197"
      Environment="ETCD_PROXY=readonly"
      Environment="ETCD_INITIAL_CLUSTER_TOKEN=etcd-cluster1"
      Environment="ETCD_DATA_DIR=/var/lib/etcd2/"
      Environment="ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379,http://127.0.0.1:8080"
      Environment="ETCD_INITIAL_CLUSTER=wwserver194=http://10.9.159.65:2380,wwserver195=http://10.9.159.66:2380,wwserver196=http://10.9.159.67:2380,wwserver197=http://10.9.159.68:2380"
     
      ExecStart=/opt/bin/etcd2
      Restart=always
      RestartSec=10
      [Install]
      WantedBy=multi-user.target

```

**Comment by bww** [comment]
Same issue here, same setup: etcd 2.0.10, Ubuntu 14.04, 3-node cluster with 1 proxy node, making a single request results in "too many open files" errors from the proxy.

**Comment by xiang90** [comment]
@bww The proxy issue has been fixed in 2.0.11

**Comment by yichengq** [comment]
This issue is a mixture, and most of them have been solved. Here are the left part:
1. If this happens on etcd 0.4, please use the latest etcd version
2. etcd 2.1 should not have this issue until cluster info used by proxy is set wrong(https://github.com/coreos/etcd/issues/2772), or there are too many clients(https://github.com/coreos/etcd/issues/2477). We will improve them.

Feel free to open a new issue if you met new case.
